lib/semantics/resolve-names.cc

```diff
@@ -3886,12 +3886,78 @@ ParamValue DeclarationVisitor::GetParamValue(const parser::TypeParamValue &x) {
 
 // ConstructVisitor implementation
 
-bool ConstructVisitor::Pre(const parser::ConcurrentHeader &) {
-  BeginDeclTypeSpec();
-  return true;
+void ConstructVisitor::ResolveIndexName(const parser::Name &name) {
+  auto *prev{FindSymbol(name)};
+  if (prev) {
+    if (prev->owner().kind() == Scope::Kind::Forall ||
+        prev->owner() == currScope()) {
+      SayAlreadyDeclared(name, *prev);
+      return;
+    }
+    name.symbol = nullptr;
+  }
+  auto &symbol{DeclareObjectEntity(name, {})};
+
+  if (symbol.GetType()) {
+    // type came from explicit type-spec
+  } else if (!prev) {
+    ApplyImplicitRules(symbol);
+  } else if (!prev->has<ObjectEntityDetails>() && !prev->has<EntityDetails>()) {
+    Say2(name, "Index name '%s' conflicts with existing identifier"_err_en_US,
+        *prev, "Previous declaration of '%s'"_en_US);
+    return;
+  } else {
+    if (auto *type{prev->GetType()}) {
+      symbol.SetType(*type);
+    }
+    if (prev->IsObjectArray()) {
+      SayWithDecl(name, *prev, "Index variable '%s' is not scalar"_err_en_US);
+      return;
+    }
+  }
+  EvaluateExpr(parser::Scalar{parser::Integer{common::Clone(name)}});
 }
-void ConstructVisitor::Post(const parser::ConcurrentHeader &) {
+
+void ConstructVisitor::ResolveControlExpression(
+    const parser::ConcurrentControl &control) {
+  Walk(std::get<1>(control.t));
+  Walk(std::get<2>(control.t));
+  const auto &optionalStep{std::get<3>(control.t)};
+  if (optionalStep.has_value()) {
+    Walk(optionalStep.value());
+  }
+}
+
+// We need to make sure that all of the index-names get declared before the
+// expressions in the loop control are evaluated so that references to the
+// index-names in the expressions are correctly detected.
+bool ConstructVisitor::Pre(const parser::ConcurrentHeader &header) {
+  BeginDeclTypeSpec();
+
+  // Process the type spec, if present
+  auto &typeSpec{std::get<std::optional<parser::IntegerTypeSpec>>(header.t)};
+  if (typeSpec.has_value()) {
+    SetDeclTypeSpec(MakeNumericType(TypeCategory::Integer, typeSpec->v));
+  }
+
+  // Process the index-name nodes in the ConcurrentControl nodes
+  auto &controls{std::get<std::list<parser::ConcurrentControl>>(header.t)};
+  for (auto &control : controls) {
+    ResolveIndexName(std::get<parser::Name>(control.t));
+  }
+
+  // Process the expressions in ConcurrentControls
+  for (auto &control : controls) {
+    ResolveControlExpression(control);
+  }
+
+  // Resolve the names in the scalar-mask-expr, if present
+  auto &maskExpr{std::get<std::optional<parser::ScalarLogicalExpr>>(header.t)};
+  if (maskExpr.has_value()) {
+    Walk(maskExpr.value());
+  }
   EndDeclTypeSpec();
+  return false;
```

Comment:
Since you're stopping at `ConcurrentHeader`, it must be the case there is no other constraints that need to be checked concerning the syntactic substructure, correct?

Reply:
@schweitzpgi, I'm not sure whether or not that's the case.  I'm just plugging through all of the constraints one at a time fixing problems and making sure that they're checked and tested.  I might be adding more checks when I deal with additional constraints.  The reason that I chose ```ConcurrentHeader``` is that ```ConcurrentHeader``` defines the scope for the ```index-name```s of the DO CONCURRENT.

Reply:
It appeared that this recursed into instances of `ConcurrentHeader` previously, but `return false;` would eliminate that recursion. It's possible that is fine. I don't know.

Reply:
I believe that returning ```false``` is OK.  My new code explicitly walks the grammatical non-terminals of a ```ConcurrentHeader```.  Let me know if there's a test case that is not covered by this code.